src/android/android_interface.rs

```diff
@@ -1 +1,155 @@
+    info!("Bootstrapping with db path: {:?}", db_path_str);
+    let db_result = db_path_str.and_then(|path| init_db(path).map_err(ServicesError::from));
+    info!("Bootstrapping result: {:?}", db_result);
+
+    jni_void_result(1, None, &env)
+}
+
+fn init_log(env: &JNIEnv, level_j_string: JString, coepi_only: jboolean, callback: jobject) -> i32 {
+    let callback_wrapper = LogCallbackWrapperImpl {
+        java_vm: env.get_java_vm().unwrap(),
+        callback: env.new_global_ref(callback).unwrap(),
+    };
+    register_callback_internal(Box::new(callback_wrapper));
+
+    let level_java_str = env.get_string(level_j_string).unwrap();
+    let level_str = level_java_str.to_str().unwrap();
+    let filter_level = LevelFilter::from_str(&level_str).expect("Incorrect log level selected!");
+    let _ = simple_logger::setup_logger(filter_level, coepi_only != 0);
+    log::max_level() as i32
+}
+
+pub fn jni_void_result(status: i32, message: Option<&str>, env: &JNIEnv) -> jobject {
+    let cls = env.find_class("org/coepi/android/api/JniVoidResult");
+
+    let status_j_value = JValue::from(status);
+
+    let msg = message.unwrap_or("");
+    let msg_j_string = env.new_string(msg).unwrap();
+    let msg_j_value = JValue::from(msg_j_string);
+
+    let obj = env.new_object(
+        cls.unwrap(),
+        "(ILjava/lang/String;)V",
+        &[status_j_value, msg_j_value],
+    );
+
+    obj.unwrap().into_inner()
+}
+
+pub fn jni_obj_result(
+    status: i32,
+    message: Option<&str>,
+    obj: JObject,
+    outer_class: &str,
+    inner_class: &str,
+    env: &JNIEnv,
+) -> jobject {
+    let cls = env.find_class(outer_class).unwrap();
+
+    let status_j_value = JValue::from(status);
+
+    let msg = message.unwrap_or("");
+
+    let msg_j_string = env.new_string(msg).unwrap();
+    let msg_j_value = JValue::from(msg_j_string);
+
+    let obj = env.new_object(
+        cls,
+        format!("(ILjava/lang/String;{})V", inner_class),
+        &[status_j_value, msg_j_value, JValue::from(obj)],
+    );
+
+    obj.unwrap().into_inner()
+}
+
+trait LogCallbackWrapper {
+    fn call(&self, level: CoreLogLevel, text: String);
+}
+
+struct LogCallbackWrapperImpl {
+    // The callback passed from Android is a local reference: only valid during the method call.
+    // To store it, we need to put it in a global reference.
+    // See https://developer.android.com/training/articles/perf-jni#local-and-global-references
+    callback: GlobalRef,
+
+    // We need JNIEnv to call the callback.
+    // JNIEnv is valid only in the same thread, so we have to store the vm instead, and use it to get
+    // a JNIEnv for the current thread.
+    // See https://developer.android.com/training/articles/perf-jni#javavm-and-jnienvb
+    java_vm: JavaVM,
+}
+
+impl LogCallbackWrapper for LogCallbackWrapperImpl {
+    fn call(&self, level: CoreLogLevel, text: String) {
+        let env = self.java_vm.attach_current_thread().unwrap();
```

Comment:
Multiple threads can be attached to VM?

Reply:
This returns a `JNIEnv` for the current thread. `JNIEnv` is valid only in the thread where it's created. Not sure about the internals.